Token counting

csibbitt opened the pull request:
## Description

* Returns token usage data in query and streaming query responses
* llama-stack currently doesn't include usage in Agent API
  * We estimate locally with an embedding model
  * Accuracy will vary, but it should be nearly 100% for GPT models.

## Type of change

- [ ] Refactor
- [x] New feature
- [ ] Bug fix
- [ ] CVE fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Configuration Update
- [ ] Bump-up service version
- [ ] Bump-up dependent library
- [ ] Bump-up library or tool used for development (does not change the final image)
- [ ] CI configuration change
- [ ] Konflux configuration change
- [ ] Unit tests improvement
- [ ] Integration tests improvement
- [ ] End to end tests improvement


## Related Tickets & Documents

- Related Issue OSPRH-17603
- Closes LCORE-178

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [x] PR has passed all pre-merge test jobs.
- [x] If it is a core feature, I have added thorough tests.

## Testing

```
$ curl -X POST http://127.0.0.1:8080/v1/query -H 'accept: ap
plication/json' -H 'Content-Type: application/json' -d '{"query": "Hi"}' | jq .

{
  "conversation_id": "7c0d30b1-a0c8-4be1-9496-6827d1745ae6",
  "response": "How can I assist you today? Do you have a question, need help with something, or just want to chat?",
  "input_tokens": 18,
  "output_tokens": 24
}

$ curl -X POST http://127.0.0.1:8080/v1/query -H 'accept: application/json' -H 'Content-Type: application/json' -d '{"query": "Please give me 200 words about the beauty of summer"}' | jq .

{
  "conversation_id": "79e2c910-2fb1-4b9e-91a0-3ffe71d53f78",
  "response": "Summer is here, and with it comes the warmth, sunshine, and endless possibilities. The long days stretch out before us like a canvas, painted with hues of blue and green as far as the eye can see. The air is filled with the sweet scent of blooming flowers and the gentle chirping of birds, creating a symphony that's both soothing and invigorating.\n\nAs we bask in the sun's warm rays, our skin feels alive, radiant and refreshed. The heat may be intense, but it's also a reminder to slow down, to take time to appreciate the simple things in life. Whether you're lounging by the pool, strolling through a park, or simply sitting on a beach, summer is a time for relaxation and rejuvenation.\n\nThe colors of summer are vibrant and bold, from the bright yellows and oranges of sunflowers to the deep blues and greens of the ocean. The warmth of the sun seeps into our souls, filling us with a sense of freedom and joy. As we soak up the sunshine, we're reminded that life is short, but it's also full of beauty and wonder. So let's make the most of this season, and let the warmth of summer fill our hearts with happiness!",
  "input_tokens": 28,
  "output_tokens": 252
}
```

I have not tested the streaming api yet, but am working on that now. I wanted to get this published for early feedback.

csibbitt commented:
> Therefore the prompt sent to the LLM by `llama-stack` will include `system_prompt`, `query_request.query` **and** the previous queries relating to the session/conversation. This is further compounded if, for example, RAG or MCP Tool Calls are included in the prompt sent to the LLM by `llama-stack`.
> I suspect the more User messages in a conversation, and the use of RAG and MCP, the more inaccurate the approach in this PR becomes.

Ah yes, I expect I will see this when I look at it using a front-end. :)
 
> Do you know if `llama-stack` have plans to support Token counting?

I could not find any public open issues related to this, but our plan was to propose it.

> Their `InferenceApi.completion` returns `CompletionResponse` that includes Token metrics however it seems this currently gets _lost_ when using an `Agent` 😢

Yes; I initially thought this would be trivial because I'd seen the usage states in the Inference API.

> It might be worth finding out their plans and keep this as a temporary solution.

This is 100% expected to be a temporary solution to get us closer to parity with road-core. You can review the internal discussion linked from OSPRH-17603 where this is framed more clearly as a stop-gap.

tisnik commented on lightspeed-stack.yaml:
```diff
@@ -22,3 +22,4 @@ user_data_collection:
   transcripts_storage: "/tmp/data/transcripts"
 authentication:
   module: "noop"
+default_estimation_tokenizer: "cl100k_base"
```
prob. worth to setup it in `constants.py` and allow to override it by anything else in `customization` block? WDYT?

tisnik commented on src/app/endpoints/query.py:
```diff
@@ -187,13 +193,21 @@ def select_model_id(models: ModelListResponse, query_request: QueryRequest) -> s
     return model_id
 
 
+def _build_toolgroups(client: LlamaStackClient) -> list[Toolgroup] | None:
```
is it part of this PR?

csibbitt replied:
It was to alleviate a "too many local variables" error that appeared while I was fixing merge conflicts.

tisnik commented on src/models/config.py:
```diff
@@ -154,6 +154,7 @@ class Configuration(BaseModel):
         AuthenticationConfiguration()
     )
     customization: Optional[Customization] = None
+    default_estimation_tokenizer: str = "cl100k_base"
```
dtto, the constant might be stored in `constants.py` with some comments what it means.

tisnik commented on src/models/responses.py:
```diff
@@ -28,10 +26,14 @@ class QueryResponse(BaseModel):
 
     conversation_id: Optional[str] = None
     response: str
+    input_tokens: Optional[int] = None
```
AGAIK it is not optional - you set `0` for any exceptions, so it's still `int` ?

tisnik reviewed (changes requested):
in overall it looks good. As @manstis wrote the actual TokenCounter implementation will be a bit more complicated, and will need to be updated when conversation history, RAG, MCP will be in place. Just some nitpicks from my side.

csibbitt commented:
I changed the status to Draft pending additions of history counting and at least some comments about where to add MCP and RAG related counting.

csibbitt commented:
I've implemented the estimation for message history as well as attachments, and addressed the other feedback. AFAICT so far, I think it would only really be possible to count tokens in RAG chunks and MCP queries from the llama-stack side; please clue me in if I've misunderstood.

Ultimately I'm not sure how useful this all is, but I'm opening it up for review so we can decide whether it's worth including or not, and so y'all can continue to point out what I've missed. :)

manstis commented on src/utils/token_counter.py:
```diff
@@ -0,0 +1,256 @@
+logger = logging.getLogger(__name__)
+
+# Class-level cache to track cumulative input tokens for each conversation
+_conversation_cache: TTLCache[str, int] = TTLCache(maxsize=1000, ttl=3600)
```
I know @eranco74 and @tisnik already have concerns about caching if there are multiple `lightspeed-stack` replicas.

Just calling this out here too...
